File: android/src/main/kotlin/com/tttocklll/watermark_kit/video/Decimation.kt
package com.tttocklll.watermark_kit.video

/**
 * Even-distribution frame decimation — single source of truth shared by the
 * surface decode loop AND the ByteBuffer fallback loop in
 * [VideoWatermarkProcessor], mirroring the pure-Dart seam
 * (`lib/src/video_decimation.dart`) and the iOS implementation so the three
 * paths cannot diverge.
 *
 * Hardcoded constants for this step (Step 4 may later parameterize via Pigeon).
 */
internal object Decimation {
  const val TARGET_FPS = 30
  const val TARGET_DURATION_SEC = 15
  const val SKIP_THRESHOLD_SEC = 15.0

  /** Synthesized output frame interval in nanoseconds (30fps). */
  const val FRAME_INTERVAL_NS = 1_000_000_000L / TARGET_FPS

  /** True when a source of [sourceDurationSec] seconds should be decimated. */
  fun shouldDecimate(
    sourceDurationSec: Double,
    skipThresholdSec: Double = SKIP_THRESHOLD_SEC
  ): Boolean = sourceDurationSec > skipThresholdSec

  /** Target output frame count (= [TARGET_FPS] * [TARGET_DURATION_SEC] = 450). */
  fun targetFrameCount(
    targetFps: Int = TARGET_FPS,
    targetDurationSec: Int = TARGET_DURATION_SEC
  ): Int = targetFps * targetDurationSec

  /**
   * Evenly-spaced source-frame indices to KEEP when decimating [totalFrames]
   * decoded frames to [target] output frames: `round(i * totalFrames / target)`
   * for `i in [0, target)`, clamped to `[0, totalFrames)`, deduped, ascending.
   * Length is `min(target, totalFrames)`. Identical math to the Dart seam.
   */
  fun selectEvenIndices(totalFrames: Int, target: Int): IntArray {
    if (totalFrames <= 0 || target <= 0) return IntArray(0)
    if (totalFrames <= target) return IntArray(totalFrames) { it }

    val kept = ArrayList<Int>(target)
    var previous = -1
    for (i in 0 until target) {
      var idx = Math.round(i.toLong() * totalFrames / target.toDouble()).toInt()
      if (idx >= totalFrames) idx = totalFrames - 1
      if (idx < 0) idx = 0
      if (idx == previous) continue
      kept.add(idx)
      previous = idx
    }
    return kept.toIntArray()
  }
}

/**
 * Per-compose decimation decision, computed once up front and shared by both
 * native decode loops.
 *
 * - [active] = false  -> keep every frame (source <= ~15s); existing behaviour.
 * - [active] = true   -> keep only the frames whose decoded index is in
 *   [keptSet]; synthesize a 30fps PTS from [DecimationState.keptIndex].
 */
internal class DecimationPlan private constructor(
  val active: Boolean,
  val totalFrames: Int,
  val target: Int,
  private val keptSet: HashSet<Int>
) {
  /** Whether the decoded frame at [decodedIndex] should be rendered/encoded. */
  fun keep(decodedIndex: Int): Boolean = !active || keptSet.contains(decodedIndex)

  companion object {
    /** Plan that keeps every frame (no decimation). */
    fun keepAll(): DecimationPlan =
      DecimationPlan(active = false, totalFrames = 0, target = 0, keptSet = HashSet())

    /**
     * Build a plan from a pre-counted [totalFrames] and the source duration.
     * Returns a keep-all plan when the source is short or the count is unusable.
     */
    fun build(totalFrames: Int, sourceDurationSec: Double): DecimationPlan {
      if (totalFrames <= 0 || !Decimation.shouldDecimate(sourceDurationSec)) {
        return keepAll()
      }
      val target = Decimation.targetFrameCount()
      if (totalFrames <= target) return keepAll()
      val indices = Decimation.selectEvenIndices(totalFrames, target)
      val set = HashSet<Int>(indices.size * 2)
      for (idx in indices) set.add(idx)
      return DecimationPlan(active = true, totalFrames = totalFrames, target = target, keptSet = set)
    }
  }
}

/** Mutable counter of kept frames, shared so synthesized PTS stays monotonic. */
internal class DecimationState {
  /** Number of frames already kept; also the synthesized-PTS frame number. */
  var keptIndex: Int = 0
    private set

  /** Synthesized 30fps presentation time in nanoseconds for the current kept frame. */
  fun nextPresentationTimeNs(): Long = keptIndex * Decimation.FRAME_INTERVAL_NS

  fun advance() { keptIndex++ }
}

File: ios/Classes/Decimation.swift
import Foundation

/// Even-distribution frame decimation — mirrors the pure-Dart seam
/// (`lib/src/video_decimation.dart`) and the Android `Decimation` object so the
/// three paths cannot diverge.
///
/// Hardcoded constants for this step (Step 4 may later parameterize via Pigeon).
enum Decimation {
  static let targetFps = 30
  static let targetDurationSec = 15
  static let skipThresholdSec: Double = 15

  /// True when a source of `sourceDurationSec` seconds should be decimated.
  static func shouldDecimate(_ sourceDurationSec: Double,
                             skipThresholdSec: Double = skipThresholdSec) -> Bool {
    return sourceDurationSec > skipThresholdSec
  }

  /// Target output frame count (= targetFps * targetDurationSec = 450).
  static func targetFrameCount(targetFps: Int = targetFps,
                               targetDurationSec: Int = targetDurationSec) -> Int {
    return targetFps * targetDurationSec
  }

  /// Evenly-spaced source-frame indices to KEEP when decimating `totalFrames`
  /// decoded frames to `target` output frames: `round(i * totalFrames / target)`
  /// for `i in [0, target)`, clamped to `[0, totalFrames)`, deduped, ascending.
  /// Length is `min(target, totalFrames)`. Identical math to the Dart seam.
  static func selectEvenIndices(totalFrames: Int, target: Int) -> [Int] {
    if totalFrames <= 0 || target <= 0 { return [] }
    if totalFrames <= target { return Array(0..<totalFrames) }

    var kept = [Int]()
    kept.reserveCapacity(target)
    var previous = -1
    for i in 0..<target {
      var idx = Int((Double(i) * Double(totalFrames) / Double(target)).rounded())
      if idx >= totalFrames { idx = totalFrames - 1 }
      if idx < 0 { idx = 0 }
      if idx == previous { continue }
      kept.append(idx)
      previous = idx
    }
    return kept
  }
}

/// Per-compose decimation decision, computed once up front.
///
/// - `active == false` -> keep every frame (source <= ~15s); existing behaviour.
/// - `active == true`  -> keep only frames whose decoded index is in `keptSet`;
///   synthesize a 30fps PTS from the kept-frame counter.
struct DecimationPlan {
  let active: Bool
  let target: Int
  private let keptSet: Set<Int>

  /// Whether the decoded frame at `decodedIndex` should be rendered/appended.
  func keep(_ decodedIndex: Int) -> Bool {
    return !active || keptSet.contains(decodedIndex)
  }

  static func keepAll() -> DecimationPlan {
    return DecimationPlan(active: false, target: 0, keptSet: [])
  }

  /// Build a plan from a pre-counted `totalFrames` and the source duration.
  static func build(totalFrames: Int, sourceDurationSec: Double) -> DecimationPlan {
    if totalFrames <= 0 || !Decimation.shouldDecimate(sourceDurationSec) {
      return keepAll()
    }
    let target = Decimation.targetFrameCount()
    if totalFrames <= target { return keepAll() }
    let indices = Decimation.selectEvenIndices(totalFrames: totalFrames, target: target)
    return DecimationPlan(active: true, target: target, keptSet: Set(indices))
  }
}
